### PR TITLE
types: fix bug which Float type is not effective in AddDate & SubDate function

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve? 
Before this pr, there are some bugs in AddDate & SubDate function like follow
```
tidb(localhost:4000) > SELECT TIME('20:20:20.1') + INTERVAL .6 SECOND;
ERROR 2013 (HY000): Lost connection to MySQL server during query

tidb(localhost:4000) > select adddate("2011-01-01 00:00:00", interval 10.1 SECOND);
+------------------------------------------------------+
| adddate("2011-01-01 00:00:00", interval 10.1 SECOND) |
+------------------------------------------------------+
| 2011-01-01 00:00:10                                  |
+------------------------------------------------------+
1 row in set (0.00 sec)
```

### What is changed and how it works?
After this pr, the queries are like follow
```
tidb(localhost:4000) > SELECT TIME('20:20:20.1') + INTERVAL .6 SECOND;
+-----------------------------------------+
| TIME('20:20:20.1') + INTERVAL .6 SECOND |
+-----------------------------------------+
| 2018-10-08 20:20:20.700000              |
+-----------------------------------------+
1 row in set (0.01 sec)

tidb(localhost:4000) > select adddate("2011-01-01 00:00:00", interval 10.1 SECOND);
+------------------------------------------------------+
| adddate("2011-01-01 00:00:00", interval 10.1 SECOND) |
+------------------------------------------------------+
| 2011-01-01 00:00:10.100000                           |
+------------------------------------------------------+
1 row in set (0.00 sec)
```
All these results are same as **MySQL**.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -2707,6 +2719,11 @@ func (c *addDateFunctionClass) getFunction(ctx sessionctx.Context, args []Expres
 			baseBuiltinFunc:      bf,
 			baseDateArithmitical: newDateArighmeticalUtil(),
 		}
+	case dateEvalTp == types.ETInt && intervalEvalTp == types.ETDecimal:
```
Comment (ngaut) [inline expression/builtin_time.go]: Could you extract a function to check if we should use Decimal math, there are some redundant code, and it's easy to make mistake to write compose equal condition.

**Comment by zz-jason** [comment]
@XuHuaiyu @tiancaiamao PTAL

File: expression/integration_test.go [hunk 2]
```diff
@@ -1797,8 +1797,8 @@ func (s *testIntegrationSuite) TestTimeBuiltin(c *C) {
 		{"\"2011-01-01 00:00:00\"", "10.10", "DAY", "2011-01-11 00:00:00", "2010-12-22 00:00:00"},
 		{"\"2011-01-01 00:00:00\"", "10.10", "HOUR", "2011-01-01 10:00:00", "2010-12-31 14:00:00"},
 		{"\"2011-01-01 00:00:00\"", "10.10", "MINUTE", "2011-01-01 00:10:00", "2010-12-31 23:50:00"},
-		{"\"2011-01-01 00:00:00\"", "10.10", "SECOND", "2011-01-01 00:00:10", "2010-12-31 23:59:50"},
-		{"\"2011-01-01 00:00:00\"", "10.10", "MICROSECOND", "2011-01-01 00:00:00.000010", "2010-12-31 23:59:59.999990"},
+		{"\"2011-01-01 00:00:00\"", "10.10", "SECOND", "2011-01-01 00:00:10.100000", "2010-12-31 23:59:49.900000"},
+		{"\"2011-01-01 00:00:00\"", "10.10", "MICROSECOND", "2011-01-01 00:00:00.000010", "2010-12-31 23:59:59.999989"},
```
Comment (tiancaiamao) [inline expression/integration_test.go]: What happens with this one?

```
mysql> select subdate('2011-01-01 00:00:00', interval 10.10 microsecond);
+------------------------------------------------------------+
| subdate('2011-01-01 00:00:00', interval 10.10 microsecond) |
+------------------------------------------------------------+
| 2010-12-31 23:59:59.999990                                 |
+------------------------------------------------------------+
1 row in set (0.00 sec)
```

Comment (lzmhhh123) [reply expression/integration_test.go]: This is a really problem. In MYSQL, it's rounding. But it round down here. I'll check it.

File: types/time.go [hunk 3]
```diff
@@ -153,7 +153,7 @@ var (
 func FromGoTime(t gotime.Time) MysqlTime {
 	year, month, day := t.Date()
 	hour, minute, second := t.Clock()
-	microsecond := t.Nanosecond() / 1000
+	microsecond := (t.Nanosecond() + 500) / 1000
```
Comment (tiancaiamao) [inline types/time.go]: It's counter-intuitive, please at least add comment here. @lzmhhh123

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**